### PR TITLE
Halo circular velocity function from Maller & Bullock 2004

### PR DESCRIPTION
**Pull request by sarahbridle**
## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/master/CONTRIBUTING.md)
- [ ] Write unit tests
- [x] Write documentation strings
- [x] Assign someone from your working team to review this pull request
- [x] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

**Review by rrjbca — CHANGES_REQUESTED** [review]
Great start! A few suggestions to improve the code:
- We prefer to avoid abbreviations in variable names where practical, so `M` should probably be renamed `mass`
- We try to use astropy units where appropriate. So you should do `from astropy.units import Unit` and then e.g. `96.6 * Unit('km s-1')` and `(mass.to_value('Msun') / 1e11)`
- Use `cosmology.h` rather than `cosmology.H0 / 100`. This also has the correct units.
- We need a test for the function. See e.g. [test_carroll](https://github.com/skypyproject/skypy/blob/master/skypy/power_spectrum/tests/test_growth.py) for a simple example of writing tests. Create a new file `skypy/halo/tests/test_properties.py` and write a function `test_circular_velocity`. An easy test would be to choose the arguments such that each bracket in the function equals one and check that the returned value equals 96.6 km/s. Perhaps you can think of others
- There are some trivial formatting issues relating to whitespace/tabs/blank lines etc. At the bottom of the pull request next to "codeclimate" click "details" to see the suggestions

File: skypy/halo/properties.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    >>> redshift = np.arange(0.3, 1, 0.5)
+    >>> properties.halo_circular_velocity(M, Delta_v, redshift, cosmology)
+    <Quantity [ 6.11303684, 36.72661831] km2 / (Mpc2 s2)>
+
+    References
+    ----------
+    .. [1] Maller and Bullock 2004 MNRAS 355 694 DOI:10.1111/j.1365-2966.2004.08349.x
+    
+	Notes of things to ask / think about:
+	* where to get Delta_v from - do we have a module for this already, or compute it in here too?
```
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:66]: we do not have a module for this yet. We could write its own function if we would reuse it or leave it as a free parameter.

File: skypy/halo/properties.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	* fix h hack / check h in Maller & Bullock is h100
+	* should we call it the circular_velocity or the virial_velocity?
```
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:68]: I would call it as the authors do, virial velocity, to avoid confusion.

File: skypy/halo/properties.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import numpy as np
+
+__all__ = [
+    'halo_circular_velocity',
+ ]
+
+def halo_circular_velocity(M, Delta_v, redshift, cosmology):
+    """Halo circular velocity.
+    This function computes the halo circular velocity, setting it 
+    equal to the virial velocity using equation (3) from [1]_.
```
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:25]:
```suggestion
    equal to the virial velocity using equation (3) and footnote 2 from [1]_.
```

File: skypy/halo/properties.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    References
+    ----------
+    .. [1] Maller and Bullock 2004 MNRAS 355 694 DOI:10.1111/j.1365-2966.2004.08349.x
+    
+	Notes of things to ask / think about:
+	* where to get Delta_v from - do we have a module for this already, or compute it in here too?
+	* fix h hack / check h in Maller & Bullock is h100
+	* should we call it the circular_velocity or the virial_velocity?
+	* ditto M - should we call it M_v or pretend virial mass = halo mass throughout?
```
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:69]: My preference would be `M_v`, as in the paper.

File: skypy/halo/properties.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	* should we output the virial radius here as well or is this already done elsewhere?
+    """
+
+    h = cosmology.H0 / 100
+    circular_velocity = 96.6 * (Delta_v * cosmology.Om0 * h **2) * pow((1 + redshift)/0.3,0.5) * pow(M/1e11,1/3)
```
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:74]: Does this correspond to the equation in footnote 2?

**Review by Lucia-Fonseca — CHANGES_REQUESTED** [review]
It needs some changes and tests.
Please, take this review as suggestions.

**Comment by Lucia-Fonseca** [comment]
We should update the description to link with the correspondent issue.

File: skypy/halo/properties.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    >>> cosmology = Planck15
+    >>> M = 10**np.arange(9.0, 12.0, 2)
+    >>> Delta_v = np.arange(1.0, 1.1, 0.1)
+    >>> redshift = np.arange(0.3, 1, 0.5)
+    >>> properties.halo_circular_velocity(M, Delta_v, redshift, cosmology)
+    <Quantity [ 6.11303684, 36.72661831] km2 / (Mpc2 s2)>
+
+    References
+    ----------
+    .. [1] Maller and Bullock 2004 MNRAS 355 694 DOI:10.1111/j.1365-2966.2004.08349.x
```
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:63]: I believe the equation comes from this other paper
https://arxiv.org/pdf/1403.1873.pdf

File: skypy/halo/properties.py (hunk 5 above)
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:74]: Perhaps I am confusing the equation but I believe it should be
```suggestion
    circular_velocity = 96.6 * np.power(Delta_v * cosmology.Om0 * h **2 / 24.4, 1.0 / 6.0) * np.sqrt((1 + redshift) / 3.3) * np.power(M / 1.0e11 , 1.0 / 3.0)
```

File: skypy/halo/properties.py (hunk 1 above)
Comment (sarahbridle) [reply skypy/halo/properties.py:65]: Great, thanks - lets make a separate function so we can call that independently if we want to (and sub in different versions of the Delta_v code)

File: skypy/halo/properties.py (hunk 4 above)
Comment (sarahbridle) [reply skypy/halo/properties.py:68]: Great, thanks. We discussed to go with halo_virial_mass !

File: skypy/halo/properties.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    ----------
+    M : (nm,) array_like
+    	Array for the virial mass, in units of solar mass.
+	Delta_v : (nm,) array_like
+		The mean overdensity of the halo.
+	redshift : (nm,) array_like
+		The redshift of each halo.
+	cosmology : astropy.cosmology.Cosmology
+        Cosmology object providing methods for the evolution history of
+        omega_matter and omega_lambda with redshift.
```
Comment (Lucia-Fonseca) [inline skypy/halo/properties.py:37]: Over indented block

File: skypy/halo/properties.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    References
+    ----------
+    .. [1] Maller and Bullock 2004 MNRAS 355 694 DOI:10.1111/j.1365-2966.2004.08349.x
+    
+	Notes of things to ask / think about:
+	* where to get Delta_v from - do we have a module for this already, or compute it in here too?
+	* fix h hack / check h in Maller & Bullock is h100
+	* should we call it the circular_velocity or the virial_velocity?
+	* ditto M - should we call it M_v or pretend virial mass = halo mass throughout?
+	* should we output the virial radius here as well or is this already done elsewhere?
```
[the remark this answers is not in the document]
Comment (sarahbridle) [reply skypy/halo/properties.py:69]: Good plan. Will make a new function, thanks.

File: skypy/halo/properties.py (hunk 2 above)
Comment (sarahbridle) [reply skypy/halo/properties.py:67]: will do!

File: skypy/halo/properties.py (hunk 7 above)
Comment (sarahbridle) [reply skypy/halo/properties.py]: hopefully now fixed thanks!

File: skypy/halo/properties.py (hunk 6 above)
Comment (sarahbridle) [reply skypy/halo/properties.py:62]: Thanks! I need to look into this carefully!

Comment (sarahbridle) [reply skypy/halo/properties.py:62]: thanks, you're right - fixed!

File: skypy/halo/properties.py (hunk 5 above)
Comment (sarahbridle) [reply skypy/halo/properties.py]: argggh, you're totally right, thank you!

**Comment by itrharrison** [comment]
Replaced with #529 to change target branch.

**Pull request closed without merging**